hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/StateContext.java

```diff
@@ -72,7 +101,13 @@
   private final AtomicLong stateExecutionCount;
   private final ConfigurationSource conf;
   private final Set<InetSocketAddress> endpoints;
-  private final Map<InetSocketAddress, List<GeneratedMessage>> reports;
+  // Only the latest full report of each type is kept
+  private final AtomicReference<GeneratedMessage> containerReports;
```

Comment:
You may make them to be `volatile`, which is simpler but thread safe as well,

Reply:
Thanks for the tip. I will keep atomic for now as the report is not updated so frequently on DNs (low performance impact).